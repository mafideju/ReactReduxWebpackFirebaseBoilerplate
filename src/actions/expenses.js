import uuid from 'uuid';
import database from '../firebase/firebase';

// FLUXO DE DADOS SEM FIREBASE
// 01: COMPONENTE CHAMA O ACTION GENERATOR
// 02: ACTION GENERATOR RETORNA UM OBJETO
// 03: COMPONENTE DISPACHA A AÇÃO ATRAVES DO MAP-DISPATCH-TO-PROPS
// 04: REDUX GUARDA AS MUDANÇAS

// FLUXO DE DADOS COM FIREBASE
// 01: COMPONENTE CHAMA O ACTION GENERATOR
// 02: ACTION GENERATOR RETORNA UMA FUNÇÃO
// 03: COMPONENTE DISPACHA FUNÇÃO (REDUX THUNK)
// 04: FUNÇÃO DISPARA UMA AÇÃO QUALQUER PREDETERMINADA

// 01: CRIAR AS ACTIONS PARA CADA AÇÃO DO USUÁRIO (COM IDENTIFICADOR UNICO)
// ADD_EXPENSE
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

// CONSIDERAÇÕES:
// 01: DATABASE UPDATE COM SUCESSO
// 02: DISPATCH O ACTION GENERATOR CORRETO
export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editExpense(id, updates));
      });
  };
};

// SET_EXPENSES
export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses`)
      .once('value')
      .then(snapshot => {
        const expenses = [];
        snapshot.forEach(snap => {
          expenses.push({
            id: snap.key,
            ...snap.val()
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
