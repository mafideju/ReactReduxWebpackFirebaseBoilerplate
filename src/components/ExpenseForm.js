import 'react-dates/initialize';
import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// const date = new Date();
// const now = moment();
// console.log(now.format('DD MMM YYYY H:mm'));

class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Complete os Campos do Formulário para Validar a Requisição'
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        <form className="form" onSubmit={this.onSubmit}>
          {this.state.error && (
            <h4 className="form__error">{this.state.error}</h4>
          )}
          <input
            type="text"
            className="text-input"
            placeholder="Descrição"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            className="text-input"
            placeholder="Valor"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            displayFormat={'DD/MM/YYYY'}
            isOutsideRange={() => false}
          />

          <textarea
            className="text-area"
            placeholder="Detalhes da Despesa"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <div>
            <button className="button button--expense button--expense__edit">
              Adicionar Despesa
            </button>
          </div>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
