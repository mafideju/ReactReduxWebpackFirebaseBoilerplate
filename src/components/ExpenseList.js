import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import { Link } from 'react-router-dom';
import ExpenseListItem from './ExpenseListItem';

export const ExpenseList = props => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Despesa</div>
      <div className="show-for-desktop">Despesa</div>
      <div className="show-for-desktop">Quantia</div>
    </div>
    {props.expenses.length === 0 ? (
      <h3 className="form__error">
        Adicione Uma Despesa para come√ßar a{' '}
        <Link to="/create">Controlar Seus Gastos</Link>
      </h3>
    ) : (
      props.expenses.map(expense => {
        return <ExpenseListItem key={expense.id} {...expense} />;
      })
    )}
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
