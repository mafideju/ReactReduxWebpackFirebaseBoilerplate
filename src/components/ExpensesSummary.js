import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import GetExpensesTotal from '../selectors/GetExpensesTotal';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const pluralItem = expenseCount === 1 ? 'item' : 'items';
  const pluralDividido = expenseCount === 1 ? 'em apenas' : 'divididos em';
  const totalGastos = numeral(expensesTotal / 100).format('$0,0.00');
  return (
    <div className="page-header">
      <div className="content-container">
        <h2 className="page-header__title">
          Total de{' '}
          <span className="page-header__title--highlight">R{totalGastos}</span>{' '}
          Reais em gastos {pluralDividido}{' '}
          <span className="page-header__title--highlight">{expenseCount}</span>{' '}
          {pluralItem}.
        </h2>
        <div>
          <Link
            className="button button--animated__left button--expense"
            to="/create"
          >
            Novo Gasto
          </Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: GetExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
