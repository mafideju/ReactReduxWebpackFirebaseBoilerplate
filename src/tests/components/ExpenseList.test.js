import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixture/expenses';

test('EXPENSELIST COM EXPENSES', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test('EXPENSELIST SEM EXPENSES', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
