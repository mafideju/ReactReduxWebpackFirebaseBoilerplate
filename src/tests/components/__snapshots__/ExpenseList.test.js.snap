// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EXPENSELIST COM EXPENSES 1`] = `
<div>
  <ExpenseListItem
    amount={1200}
    createdAt={0}
    description="Aluguel"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={550}
    createdAt={-345600000}
    description="Mercado"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={280}
    createdAt={345600000}
    description="Contas de Consumo"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`EXPENSELIST SEM EXPENSES 1`] = `
<div>
  <h3>
    Adicione Uma Despesa para 
    <Link
      replace={false}
      to="/create"
    >
      Come√ßar sua Carteira
    </Link>
  </h3>
</div>
`;
